File: scrape_tipas.R
# Sitio: Las Tipas
# Autor: Joaquin Bermejo
# Fecha: Marzo 8, 2025
# ------------------------------------------
#  1. Acceder a la pagina
#  2. Para cada fecha:
#      a. Ir a la fecha
#      b. Obtener las peliculas en cartelera
#      c. Para cada pelicula disponible:
#          I. Obtener el nombre y formato
#          II. Para cada horario, extraer hora e idioma y grabar

library(RSelenium)
library(rvest)
library(netstat)
library(tidyverse)

remote_driver <- rsDriver(browser = "firefox", port = free_port(), verbose = F, phantomver = NULL)
remDr <- remote_driver$client

#  1. Acceder a la pagina
remDr$navigate("https://rosario.lastipas.com.ar/es-AR")
Sys.sleep(5)

btns_fechas <- remDr$findElements("css selector", "#snapDateSlider > div")
btn_derecha <- remDr$findElements("css selector", ".gap-4 > button")[[2]]
fechas_texto <- unlist(lapply(btns_fechas, function(x) x$getElementAttribute("textContent")[[1]]))
fechas <- character()
meses_espanol <- c(
  "Enero"=1, "Febrero"=2, "Marzo"=3, "Abril"=4, "Mayo"=5, "Junio"=6,
  "Julio"=7, "Agosto"=8, "Septiembre"=9, "Octubre"=10, "Noviembre"=11, "Diciembre"=12
)
for (fecha_texto in fechas_texto) {
  dia <- as.numeric(str_extract(fecha_texto, "\\d+"))
  mes <- meses_espanol[[str_extract(fecha_texto, "[A-Za-z]+$")]]
  fecha <- as.character(as.Date(paste(year(Sys.Date()), mes, dia, sep = "-")))
  fechas <- c(fechas, fecha)
}

data <- data.frame(Nombre = NULL, Fecha = NULL, Tipo = NULL, Horario = NULL)

#  2. Para cada fecha:
for (i in 1:length(fechas)) {
  
  #      a. Ir a la fecha
  btns_fechas[[i]]$clickElement()
  btn_derecha$clickElement()
  Sys.sleep(1)
  
  #      b. Obtener las peliculas en cartelera
  pelis <- remDr$findElements("class name", "max-w-[670px]")
  
  #      c. Para cada pelicula disponible:
  for (peli in pelis) {
    #          I. Obtener el nombre y formato
    nombre_peli <- peli$findChildElement("css selector", "h4")$getElementText()[[1]]
    nombre_peli <- sub("(.*?)\\s*(2D|3D|ATMOS).*", "\\1", nombre_peli) # remover 2D, 3D, 4D
    formato <- peli$findChildElements("css selector", "p")[[8]]$getElementText()[[1]]
    
    #          II. Para cada horario, extraer hora e idioma y grabar
    horarios_spans <- peli$findChildElements("css selector", "span.text-primary.flex")
    horarios <- unlist(lapply(horarios_spans, function(x) x$getElementText()[[1]]))
    for (horario in horarios) {
      horario_split <- unlist(strsplit(horario, "\n"))
      tipo <- ifelse(length(horario_split) > 1, paste(formato, horario_split[1]), formato)
      hora <- horario_split[length(horario_split)]
      data <- rbind(data, data.frame(Nombre = nombre_peli, Fecha = fechas[i], Tipo = tipo, Horario = hora))
    }
    
  }
  
}

remDr$close()

write.csv(data, "pelis_tipas.csv", row.names = F, fileEncoding = "UTF-8")

File: scrape_tipas.py
from selenium import webdriver
from selenium.webdriver.firefox.options import Options
from selenium.webdriver.common.by import By
import pandas as pd
import time
import re
from datetime import datetime

# Configure headless Firefox
options = Options()
options.headless = True
driver = webdriver.Firefox(options=options)

# 1. Go to the website
driver.get("https://rosario.lastipas.com.ar/es-AR")
time.sleep(5)

# Get date buttons and extract text
btns_fechas = driver.find_elements(By.CSS_SELECTOR, "#snapDateSlider > div")
btn_derecha = driver.find_elements(By.CSS_SELECTOR, ".gap-4 > button")[1]  # Click right arrow after each selection

meses_es = {
    "Enero": 1, "Febrero": 2, "Marzo": 3, "Abril": 4, "Mayo": 5, "Junio": 6,
    "Julio": 7, "Agosto": 8, "Septiembre": 9, "Octubre": 10, "Noviembre": 11, "Diciembre": 12
}

fechas = []
btns_actuales = []

for btn in btns_fechas:
    texto = btn.text.strip()
    match = re.search(r"(\d+)\s+([A-Za-z]+)", texto)
    if match:
        dia = int(match.group(1))
        mes = meses_es.get(match.group(2), 1)
        fecha = datetime(datetime.today().year, mes, dia).strftime("%Y-%m-%d")
        fechas.append(fecha)
        btns_actuales.append(btn)

data = []

# 2. Iterate through each date
print(fechas)
for i, fecha in enumerate(fechas):
    try:
        btns_actuales[i].click()
        btn_derecha.click()  # simulate swiping
        time.sleep(1)

        pelis = driver.find_elements(By.XPATH, "//div[contains(@class, 'max-w-')]")

        for peli in pelis:
            try:
                # I. Movie name and format
                nombre_raw = peli.find_element(By.CSS_SELECTOR, "h4").text
                nombre = re.sub(r"\s*(2D|3D|ATMOS|4D).*", "", nombre_raw).strip()

                formato = peli.find_elements(By.CSS_SELECTOR, "p")[6].text  # 7th <p> tag

                # II. Get showtimes
                horarios = peli.find_elements(By.CSS_SELECTOR, "span.text-primary.flex")
                for h in horarios:
                    contenido = h.text.strip().split('\n')
                    if len(contenido) == 2:
                        idioma, hora = contenido
                        tipo = f"{formato} {idioma}".strip()
                    else:
                        hora = contenido[0]
                        tipo = formato
                    data.append({"Nombre": nombre, "Fecha": fecha, "Tipo": tipo, "Horario": hora})
            except Exception as e:
                print(f"Error processing movie: {e}")
    except Exception as e:
        print(f"Error on date {fecha}: {e}")

# Save to CSV
df = pd.DataFrame(data)
df.to_csv("pelis_tipas.csv", index=False, encoding="utf-8")

driver.quit()
